subprojects/client-protocol/src/main/java/org/gradle/profiler/client/protocol/serialization/MessageSerializer.java

```diff
@@ -1,57 +1,164 @@
+            return new StudioRequest(syncId, requestType);
         }
+    },
+    STUDIO_SYNC_REQUEST((byte) 6, StudioSyncRequestCompleted.class) {
+        @Override
+        public void doWriteTo(Connection connection, Message message) throws IOException {
+            StudioSyncRequestCompleted request = (StudioSyncRequestCompleted) message;
+            connection.writeInt(request.getId());
+            connection.writeLong(request.getDurationMillis());
+            connection.writeString(request.getResult().toString());
+        }
+
+        @Override
+        public Message doReadFrom(Connection connection) throws IOException {
+            int syncRequestCompletedId = connection.readInt();
+            long syncRequestCompletedDurationMillis = connection.readLong();
+            StudioSyncRequestResult result = StudioSyncRequestResult.valueOf(connection.readString());
+            return new StudioSyncRequestCompleted(syncRequestCompletedId, syncRequestCompletedDurationMillis, result);
+        }
+    }
+    ;
+
+    private static final Map<Class<? extends Message>, MessageSerializer> SERIALIZERS_BY_CLASS = Collections.unmodifiableMap(
+        Arrays.stream(values()).collect(Collectors.toMap(e -> e.type, Function.identity())));
+
+    private static final Map<Byte, MessageSerializer> SERIALIZERS_BY_MESSAGE_ID = Collections.unmodifiableMap(
+        Arrays.stream(values()).collect(Collectors.toMap(e -> e.messageId, Function.identity())));
+
+    private final byte messageId;
+    private final Class<? extends Message> type;
+
+    MessageSerializer(byte id, Class<? extends Message> type) {
+        this.messageId = id;
+        this.type = type;
+    }
+
+    protected abstract Message doReadFrom(Connection connection) throws IOException;
+    protected abstract void doWriteTo(Connection connection, Message message) throws IOException;
+
+    /**
+     * Reads a message from the given connection.
+     */
+    public Message readFrom(Connection connection) throws IOException {
+        return doReadFrom(connection);
     }
 
-    public <T extends Message> T receive(Class<T> type, Duration timeout) {
-        try {
-            Future<Object> future = executorService.submit((Callable<Object>) this::receive);
-            Object result = future.get(timeout.toMillis(), TimeUnit.MILLISECONDS);
-            return type.cast(result);
-        } catch (TimeoutException e) {
-            throw new IllegalStateException(String.format("Timeout waiting to receive message from %s.", peerName));
-        } catch (ExecutionException | InterruptedException e) {
-            throw new RuntimeException(e);
+    /**
+     * Writes a message to the given connection.
+     */
+    public void writeTo(Connection connection, Message message) throws IOException {
+        connection.writeByte(messageId);
+        doWriteTo(connection, message);
+    }
+
+    public static MessageSerializer findMessageSerializer(Class<? extends Message> messageClass) {
```

Comment:
I'd say these methods should be called `get...()` — we typically use `find...()` for getters that are not expected to always find a single instance.

Reply:
fixed